src/cargo/core/compiler/mod.rs

```diff
@@ -999,3 +987,33 @@ impl Kind {
         }
     }
 }
+
+fn json_stdout(line: &str) -> CargoResult<()> {
+    if !line.is_empty() {
+        Err(internal(&format!(
+            "compiler stdout is not empty: `{}`",
+            line
+        )))
+    } else {
+        Ok(())
+    }
+}
+
+fn json_stderr(line: &str, package_id: &PackageId, target: &Target) -> CargoResult<()> {
+    // stderr from rustc/rustdoc can have a mix of JSON and non-JSON output
```

Comment:
Hm, does `rustdoc` put JSON on `stderr` as well? It's a shame that Cargo and Rustc use different streams for JSON =/

Reply:
It is an odd inconsistency.  There's an open issue for that (rust-lang/rust#48301).  I imagine since the vast majority of people use rustc behind cargo it doesn't really come up often.